### ISSUE TITLE
v2.4.306 not only adds Ads but also Firebase Analytics & GMS

### ISSUE THREAD

**Issue by IzzySoft** [vpnhood/VpnHood#136]
Can you please provide a build without those? No offense meant, but a VPN solution (intended to keep your data private) that includes Tracking by one of the biggest data collectors simply is a no-go. For now I've removed that update from my repo and disabled updates until this problem is solved. Should it not be possible to have such a F/LOSS variant, I will unfortunately have to remove your app from my repo again as it no longer meets the inclusion criteria (and it makes no sense to keep an outdated version online).

What goes to Google Play is of course none of my concerns, though if you really have privacy in mind you should consider that as well. I guess Analytics sneaked in unintentionally when implementing Google Ads, so you'd probably be willing to at least disable that even for Play Store.

**Comment by trudyhood** [comment]
Hi, 

Firebase Analytics was already there from the first version.  I believe it is a good idea to make it optional for end-users but we need to implement it.

Google Ad shows only fires on public servers. We also hate Ads, but unfortunately, we need funds to add some free public servers; otherwise, it just leads us to shut down our public servers. We think having an Ad is better than nothing. The free public servers are heavily under pressure.

We are sorry that you remove VpnHood from your repo. You are welcome if you have any idea how we can fund the free public servers in any other way.

We will let you know if we find another way to fund our services and remove the Ads. Another idea is to display a red warning on the app page in your repo that the app track users with Google Ad or Analytics. Let us know if you change your policy and bring back VpnHood to your repo.

**Comment by IzzySoft** [comment]
> Firebase Analytics was already there from the first version

It didn't show up before. This new version was triggering the alert for it, the previous ones did not. GMS was already there, yes (which is also questionable and probably means the app won't run on Google-free devices; it's also already border-line concerning the inclusion criteria, see below).

> we need to implement it.

If you insist on this specific analytics package, I'll have to remove the app as it no longer meets the criteria – which I'd find a sad thing to do. Maybe [an alternative](https://gitlab.com/-/snippets/1896503#acceptable-analytics) would do if you think you really need analytics? Something opt-in preferably?

Google Ads: I've nothing against Ads per se (I don't like them much in the way they usually come, but that's a different thing). Here it's rather the "trust-able source". Google being the data collector it is is not such source. The risk is this proprietary code is siphoning private data, and thus contradicting the purpose of a VPN. If you had your own ads implementation using your own servers (even as proxies to isolate), that'd be no problem in this context.

And yes, I see the reason behind the monetization element. Which is why I try to find a way to keep your app in my repo (you have noticed that I just removed the latest update, not yet the app itself). Maybe an own library proxying ad requests could work for you (I know some apps do that but do not remember which); you could take a look at [F-Droid apps with the `Ads` anti-feature](https://monitor.f-droid.org/anti-feature/Ads) to see what they are using. I'm not aware of any FOSS ad libs we could call "safe" in this context unfortunately (should you find some please let me know and I add them to my snippet). Asking for donations would be another way, or setting up a paid plan including some goodies…

> We will let you know if we find another way to fund our services and remove the Ads.

Please do. I'll keep the current version frozen for a while before taking "further action".

> Another idea is to display a red warning on the app page in your repo that the app track users with Google Ad or Analytics.

Nope. The combination of those with apps dealing with sensitive information (VPN, children, password safes etc) is off limits, and that won't change.

Bringing in another option:build flavors. I indicated before that what you publish at Play is not in my backyard. And the majority of your users certainly installs the app from there. So you could have a `foss` flavor free of the 3 mentioned candidates (making that APK available at your releases in addition to the Play one) which goes to my repo then, while serving the other to the majority at PlayStore – while looking for alternatives. Also, adding a `FUNDING.yml` with donation possibilities could help your finances (my updater pulls that along automatically if found).

**Comment by trudyhood** [comment]
> It didn't show up before. This new version was triggering the alert for it, the previous ones did not.
Maybe your repo just looks for some library in APK, we have been using firebase javascript in its webview. Perhaps adding GoogleAds library adds some dependency to firebase so your repo checker finds it now. You may need to improve the checker such as running the app on an emulator and watching the network traffic to find it out.

We need to implement a version and isolate the Google Ad-free version, such as applying monthly usage or limiting speed for anonymous users. Another option is to remove the public server service from an APK like OpenVPN client but I think most users download the app just for public servers and they will blame the app doesn't offer any VPN service.
These need much time to implement.

> Please do. I'll keep the current version frozen for a while before taking "further action".
Our free public servers will stop supporting the old version. If we can't find enough funds or get bad feedback for adding Ad (who likes Ads!), we will have to shut them down and release a software-only app. So publishing APK with the free server without ads or failure to find any organization funding the servers is detrimental to public servers.

**Comment by IzzySoft** [comment]
> Maybe your repo just looks for some library in APK

True for my scanner, yes.

> You may need to improve the checker such as running the app on an emulator and watching the network traffic to find it out.

Haha, that would be nice, yeah – but I don't have the resources for that. I might look into integrating something like [Pithus](https://beta.pithus.org/report/ea23095447263df14eae92f3f70ad76e3caef5c58a5f6b504a9273a4ed63b8eb) in addition to the existing checkers, if they provide an API (but they also do static checks only). Or use the advanced features of [VirusTotal](https://www.virustotal.com/gui/file/ea23095447263df14eae92f3f70ad76e3caef5c58a5f6b504a9273a4ed63b8eb/behavior) (they e.g. found Google Analytics and TagManager in their behavior analysis of your app's previous version, but those APIs need a paid subscription – raising the question of who shall pay for that). I'm just a single person and must make my living as well (I don't earn any money by running my services), so I have no idea what time I should do that. I do what I can, but my resources are limited.

> These need much time to implement.

I understand the dilemma.

> Our free public servers will stop supporting the old version.

When will that be? At that point I would have to remove it (no sense in keeping non-working versions).

> If we can't find enough funds or get bad feedback for adding Ad (who likes Ads!), we will have to shut them down and release a software-only app. So publishing APK with the free server without ads or failure to find any organization funding the servers is detrimental to public servers.

Life's a bitch :cry: But I cannot list apps which do not comply – as much sympathy as I have, that defeats the purpose. If we can find a working solution, I gladly keep it up – but if we can't, I cannot.

As pointed out, I'll keep the door open as long as I can. Once it breaks, I'll have to unlist – but if at a later point you get a solution, we can always re-establish the listing within a day.

**Comment by IzzySoft** [comment]
@trudyhood one more option just came to mind: Have a popup on first start pointing out the situation in short, pregnant terms and set up a funding campaign. You could e.g. use [Liberapay](https://liberapay.com/) where people can configure their own subscriptions (like "I wish to send you USD 2 each month"), or similar platforms. Liberapay you can then also mention in you `FUNDING.yml` so it gets shown with your app listing automatically as well.

**Comment by trudyhood** [comment]
> one more option just came to mind: Have a popup on first start pointing out the situation in short, pregnant terms and set up a funding campaign. You could e.g. use [Liberapay](https://liberapay.com/) where people can configure their own subscriptions (like "I wish to send you USD 2 each month"), or similar platforms. Liberapay you can then also mention in you FUNDING.yml so it gets shown with your app listing automatically as well.

Yes, I found it today. I am not sure how many people will contribute, but we have to try it out. Thank you for sharing it with us.

**Comment by trudyhood** [comment]
> Add arrr (pirate chain) crypto donations within app and remove all shits ads, aggressive permissions (if there are) and trackers. Then people can trust of your app.

Thank you for your comments.

First, we plan to make the tracker optional so users can opt-in to share anonymous data.

Second, The App is different than our free public servers. Users can use VpnHood and use their server without any AD. Everything is free and open source, and we are here to help them. However, the Vpn Server services are another story. They will see the AD for that, we also promoted a message that they are tracked on our public server so they can cancel, but we have found that many people don't care about that level of privacy and need to bypass censorship, so it helps them a lot.

A problem remains; some checking tools see the reference of some libraries, such as Google AD or Analytics, regardless of the app's use. If you see, we initialize the GoogleAD only after the user connects to a public server.

Another idea is publishing two different versions of the VpnHood client and removing the free service from the other one.

**Comment by IzzySoft** [comment]
> I am not sure how many people will contribute, but we have to try it out.

Cannot hurt – nothing to lose. I had to give up on them once Mangopay (bank transfers) dropped them, as Paypal and Stripe are no options for me. But before that happened, my experiences were quite good. No big sums, but decent.

> Add arrr (pirate chain) crypto donations

the number of users for a given exotic crypto currency is probably even lower – but yes, putting information to where donations can happen and have multiple choices available is no bad idea. Depending on your operation model, there'd also be OpenCollective, Patreon, Kofi and several others.

> our free public servers

The "free plan" could have some "decent limits" so some nice-to-haves would be paid. Users could obtain a "key" from you to enter in their app(s) to activate the extras. Beware of "key sharing" by e.g. limiting the number of devices allowed to be online at the same time using the same key.

> we initialize the GoogleAD only after the user connects to a public server.

That is what you say, and also what you intend. But with the code being proprietary, it's nothing you can *guarantee* – you cannot analyze the code for that library as it's not available. That's what I referred to earlier with "proxying ads" (i.e. using your own lib pulling the ads from your server, which in its own name and with its own IP pulls them from wherever you have trust in; something I use with my websites to present context-matching ads from Amazon – which then only "fire" when the user clicks them).

> Another idea is publishing two different versions of the VpnHood client and removing the free service from the other one.

That could be an interim solution you could even match with the "key file" I've mentioned above: once there's a valid key file entered, the paid options are made available inside the app.

All that are just ideas collected which hopefully help you. And thanks to @serrq to not having "simply complained and dropped the app" but contributing to possible solutions :heart_eyes: That's the spirit of FOSS: I can no longer use something for reasons the dev has control over – so I try to contribute to make that possible. Chapeau! You can be proud having such users, @trudyhood – and I really hope that encourages you (of course one cannot live off words – but the spirit wants to live as well).

**Comment by bobvhood** [comment]
Hi Everyone.
We have removed the Google ADS and its library from 310. That couldn't really solve our funding problem.

Thank you for all of your contributions.

**Comment by IzzySoft** [comment]
Thanks @bobvhood – confirmed all non-free deps gone. Updates in my repo are re-enabled, v2.4.310 will become available with the next sync.

Still sad to read your funding problem wasn't solved. I hope I have read that correctly that those Ads didn't look helpful – as after just 2 days one cannot say much about things like Liberapay, OpenCollective, Kofi & Co. I'd definitely recommend giving Liberapay a go, and if conditions meet also OpenCollective. Seeing how well it works with F-Droid, and from my experiences with Liberapay back when they still supported payouts per bank transfer, that's really promising. And having a tiny nice (but not crucial) option reserved for "pro users" should be motivating them for donations as well.
